src/nvim/diff.c

```diff
@@ -1774,20 +1764,253 @@ void diff_clear(tabpage_T *tp)
+      isfiller = (curdif->df_count[toidx] ? 0 : 1);
+      if (isfiller) {
+        while (curdif && curdif->df_next && curdif->df_lnum[toidx] ==
+            curdif->df_next->df_lnum[toidx] &&
+            curdif->df_next->df_count[toidx] == 0) {
+          curdif = curdif->df_next;
+          ch_virtual_lines += get_max_diff_length(curdif);
+        }
+      }
+      if (curdif) {
+        curdif = curdif->df_next;
+      }
+    }
+  }
+}
+
+void calculate_topfill_and_topline(const int fromidx, const int toidx, const
+    int from_topline, const int from_topfill, int *topfill, linenr_T
+    *topline)
+{
+  // 1. find the position from the top of the diff block, and the start
+  // of the next diff block
+  diff_T *thistopdiff = NULL;
+  diff_T *nextblockblock = NULL;
+  int virtual_lines_passed = 0;
+
+  find_top_diff_block(&thistopdiff, &nextblockblock, fromidx, from_topline);
+
+  // count the virtual lines that have been passed
+
+  diff_T *curdif = thistopdiff;
+  while (curdif && ( curdif->df_lnum[fromidx] + curdif->df_count[fromidx] )
+      <= from_topline) {
+
+    virtual_lines_passed += get_max_diff_length(curdif);
+
+    curdif = curdif->df_next;
+  }
+
+  if (curdif != nextblockblock) {
+    virtual_lines_passed += from_topline - curdif->df_lnum[fromidx];
+  }
+  virtual_lines_passed -= from_topfill;
+
+
+  // count the same amount of virtual lines in the toidx buffer
+  curdif = thistopdiff;
+  int curlinenum_to = thistopdiff->df_lnum[toidx];
+  int linesfiller = 0;
+  count_filler_lines_and_topline(&curlinenum_to, &linesfiller,
+      thistopdiff, toidx, virtual_lines_passed);
+
+  // count the number of filler lines that would normally be above this line
+  int maxfiller = 0;
+  for (diff_T *dpfillertest = thistopdiff; dpfillertest != NULL; dpfillertest = dpfillertest->df_next) {
+    if (dpfillertest->df_lnum[toidx] == curlinenum_to) {
+      while (dpfillertest && dpfillertest->df_lnum[toidx] == curlinenum_to) {
+        maxfiller += dpfillertest->df_count[toidx] ? 0 : get_max_diff_length(dpfillertest);
+        dpfillertest = dpfillertest->df_next;
+      }
+      break;
+    }
+  }
+  (*topfill) = maxfiller - linesfiller;
+  (*topline) = curlinenum_to;
+}
+
+void run_linematch_algorithm(diff_T * dp)
+{
+  int i, j;
+  // include the algorithm and run it here
+
+  // define buffers for diff algorithm
+  diffin_T *diffbuffers = xmalloc(sizeof(diffin_T) * DB_COUNT);
+  const char **diff_begin = xmalloc(sizeof(char *) * DB_COUNT);
+  int *diff_length = xmalloc(sizeof(int) * DB_COUNT);
+  int *outputmap = xmalloc(sizeof(int) * DB_COUNT);
+  int diffbuffers_count = 0;
+  for (i = 0; i < DB_COUNT; i++) {
+    if (curtab->tp_diffbuf[i] != NULL) {
+      memset(&diffbuffers[diffbuffers_count], 0, sizeof(diffbuffers[diffbuffers_count]));
+      diff_write(curtab->tp_diffbuf[i], &diffbuffers[diffbuffers_count]);
+      diff_begin[diffbuffers_count] = diffbuffers[diffbuffers_count].din_mmfile.ptr;
```

Comment:
I have never used non internal diff. I've used all the different diff algorithms like histogram and patience, and it works fine with linematch enabled, but I assume you are referring to setting diffexpr to something? this I have never done, but from looking through the code, I don't see how this would have any effect on the linematch algorithm. 
Here, I am using the function diff_write to get access to the buffers that are being diffed this all takes place after whatever diff algorithm is going to run to create the diff hunks has ran. I used diff_write because I looked at the way the buffers are passed to to the function diff_file(diffio_T *dio), and I did the same thing.

When I say access to the buffers, I mean an array of char for each file that is diffed.

Reply:
I tested it today with using an external diff expression, and saw no issues with using the linematch feature at the same time